utils/supabase/server.ts

```diff
@@ -0,0 +1,113 @@
+import { createServerClient, type CookieOptions } from '@supabase/ssr';
+import { cookies } from 'next/headers';
+import { Database } from '@/types_db';
+import { cache } from 'react';
+
+export const createClient = (cookieStore: ReturnType<typeof cookies>) => {
+  return createServerClient<Database>(
+    process.env.NEXT_PUBLIC_SUPABASE_URL!,
+    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
+    {
+      cookies: {
+        get(name: string) {
+          return cookieStore.get(name)?.value;
+        },
+        set(name: string, value: string, options: CookieOptions) {
+          try {
+            cookieStore.set({ name, value, ...options });
+          } catch (error) {
+            // The `set` method was called from a Server Component.
+            // This can be ignored if you have middleware refreshing
+            // user sessions.
+          }
+        },
+        remove(name: string, options: CookieOptions) {
+          try {
+            cookieStore.set({ name, value: '', ...options });
+          } catch (error) {
+            // The `delete` method was called from a Server Component.
+            // This can be ignored if you have middleware refreshing
+            // user sessions.
+          }
+        }
+      }
+    }
+  );
+};
+
+const createCacheClient = cache(() => {
```

Comment:
What is the benefit of using the `cache` function here? To my knowledge, the `createClient` and `cookies` functions are quite inexpensive and synchronous, so probably don't require caching. Also, given the session inside the cookie is expected to be refreshed while the user is using the app, and the React docs specify that the cached version will be used if the function is called with the same arguments, and the `createCacheClient` is always called with no arguments. I think this could get in a state where the cookies are stale! No one likes a stale cookie! 🍪

Additionally, one of the ways that Next.js detects whether a route should be dynamic is by checking whether the `cookies` function is called from within the scope of the Server Component (or Route Handler). Since calling the `cookies` function is obscured by these functions calling `createCacheClient`, Next.js will not automatically detect that these routes should be dynamic. This causes a build "warning", and if enough of these happen, 
the build fails. Therefore, it is recommended to always call the `cookies` function from the scope of the Server Component or Route Handler - similar to how it is being done with the `createClient` function above 👍

Reply:
@dijonmusters per https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#sharing-data-between-components React's cache can be used to prevent duplicate requests being made, is there an official example of this working with Supabase somewhere?